### PR TITLE
[Test] Validate RESTful operations

### PR DESCRIPTION
**Pull request by sayaleeraut**
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/mgr/test_restful_operations_pool_get.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    "Accept": "application/vnd.ceph.api.v1.0+json",
+    "Content-Type": "application/json",
+}
+
+# Missing config parameter err msg
+CONFIG_ERR_MSG = "Mandatory config parameter '{}' not provided"
+
+
+# Mandatory test config exception
+class ConfigNotFoundError(Exception):
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_get.py:22]: Add `ConfigError` exception under https://github.com/red-hat-storage/cephci/blob/master/api/exceptions.py

**Review by vamahaja — CHANGES_REQUESTED** [review]
All test files can be merged into single file.

File: tests/mgr/test_restful_operations_pool_get.py (hunk 1 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_get.py:22]: https://github.com/red-hat-storage/cephci/blob/master/api/exceptions.py is giving 404

**Comment by sayaleeraut** [comment]
> All test files can be merged into single file.

Do you mean we don't need 3 test files ?
We discussed to have separate .py files for each method if I remember correctly

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    if not username:
+        raise ConfigNotFoundError(CONFIG_ERR_MSG.format("username"))
+
+    endpoint = config.get("endpoint")
+    if not endpoint:
+        raise ConfigNotFoundError(CONFIG_ERR_MSG.format("endpoint"))
+
+    method = config.get("method")
+    if not method:
+        raise ConfigNotFoundError(CONFIG_ERR_MSG.format("endpoint"))
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:44]: `endpoint` ?

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        if not pool_id:
+            raise OperationFailedError(f"Failed to get pool_id for {pool_name}")
+        endpoint += f"/{pool_id}"
+
+        # Use PATCH method on the given pool_id of the RESTful module
+        _code, _ = Pool(url=url, api=endpoint).patch(
+            username=username, key=key, data=data, check_sc=True
+        )
+        if _code == status_code:
+            LOG.info("RESTful API validation is successful")
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:129]: same as above

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    # Get key for API user
+    key = json.loads(CephAdm(node).ceph.restful.list_key()).get(username, None)
+    if not key:
+        raise OperationFailedError(f"Failed to get key for API user {username}")
+
+    # Declare the URL for the request
+    url = f"https://{mgr_ip}:8003"
+
+    if method == "GET":
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:58]: better to create separate methods for `get`, `post` etc methods.

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            raise OperationFailedError(
+                f"RESTful GET method for {endpoint} failed with status code:{_code}"
+            )
+    elif method == "POST":
+        data = config.get("data")
+        if not data:
+            raise ConfigNotFoundError(CONFIG_ERR_MSG.format("data"))
+
+        # Use POST method on the /pool endpoint of the RESTful module
+        url = f"https://{mgr_ip}:8003"
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:92]: variable redeclaration

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                    pool_id = pool["pool_id"]
+                    break
+            if not pool_id:
+                raise OperationFailedError(f"Failed to get pool_id for {pool_name}")
+            endpoint += f"/{pool_id}"
+        _code, _ = Pool(url=url, api=endpoint).get(
+            username=username, key=key, check_sc=True
+        )
+        if _code == status_code:
+            LOG.info("RESTful API validation is successful")
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:81]: this is not required, can be removed.

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        if not data:
+            raise ConfigNotFoundError(CONFIG_ERR_MSG.format("data"))
+
+        # Use POST method on the /pool endpoint of the RESTful module
+        url = f"https://{mgr_ip}:8003"
+        _code, _ = Pool(url=url, api=endpoint).post(
+            username=username, key=key, data=data, check_sc=True
+        )
+        if _code == status_code:
+            LOG.info("RESTful API validation is successful")
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:97]: same as above

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/mgr/test_restful_operations_pool_get.py (hunk 1 above)
Comment (vamahaja) [reply tests/mgr/test_restful_operations_pool_get.py]: create new file and add exception

File: tests/mgr/test_restful_operations_pool_endpoint.py (hunk 4 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_endpoint.py:135]: Done

File: tests/mgr/test_restful_operations_pool_endpoint.py (hunk 2 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_endpoint.py]: Corrected

File: tests/mgr/test_restful_operations_pool_get.py (hunk 1 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_get.py]: Done

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                pool_id = pool["pool_id"]
+                break
+        if not pool_id:
+            raise OperationFailedError(f"Failed to get pool_id for {pool_name}")
+        endpoint += f"/{pool_id}"
+    _code, _ = Pool(url=url, api=endpoint).get(
+        username=username, key=key, check_sc=True
+    )
+    if _code == status_code:
+        LOG.info("RESTful API validation is successful")
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:42]: this statement is not required, kindly remove.

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+def _post(config, url, endpoint, username, key, status_code):
+    data = config.get("data")
+    if not data:
+        raise ConfigError(CONFIG_ERR_MSG.format("data"))
+    # Use POST method on the /pool endpoint of the RESTful module
+    _code, _ = Pool(url=url, api=endpoint).post(
+        username=username, key=key, data=data, check_sc=True
+    )
+    if _code == status_code:
+        LOG.info("RESTful API validation is successful")
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:58]: this statement is not required, kindly remove.

File: tests/mgr/test_restful_operations_pool_endpoint.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            break
+    if not pool_id:
+        raise OperationFailedError(f"Failed to get pool_id for {pool_name}")
+    endpoint += f"/{pool_id}"
+    # Use PATCH method on the given pool_id of the RESTful module
+    _code, _ = Pool(url=url, api=endpoint).patch(
+        username=username, key=key, data=data, check_sc=True
+    )
+    if _code == status_code:
+        LOG.info("RESTful API validation is successful")
```
Comment (vamahaja) [inline tests/mgr/test_restful_operations_pool_endpoint.py:88]: this statement is not required, kindly remove.

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/mgr/test_restful_operations_pool_endpoint.py (hunk 10 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_endpoint.py]: Done, removed at all places.

File: tests/mgr/test_restful_operations_pool_endpoint.py (hunk 8 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_endpoint.py]: Done, removed.

File: tests/mgr/test_restful_operations_pool_endpoint.py (hunk 9 above)
Comment (sayaleeraut) [reply tests/mgr/test_restful_operations_pool_endpoint.py]: Done, removed.

**Review by pranavprakash20 — APPROVED** [review]
LGTM

**Review by vamahaja — APPROVED** [review]

**Pull request merged**